缩略图生成，如果使用保底方案（通过内部访问原图）生成，可能会因为 Nginx 拦截导致无法生成

### 系统信息

- 外部访问地址: https://blog.fit2cloud.com
- 启动时间: 2024-09-10 12:30
- 版本: 2.19.2
- 构建时间: 2024-09-06 22:17
- Git Commit: 7f68de3
- Java: IBM Semeru Runtime Open Edition / 21.0.4+7-LTS
- 数据库: PostgreSQL / 15.5 (Debian 15.5-1.pgdg120+1)
- 操作系统: Linux / 5.15.0-92-generic
- 已激活主题: [FIT2BLOG 1.3.0](https://www.halo.run/store/apps/app-TXBPo)
- 已启动插件:
  - [页面静态缓存 1.2.0](https://github.com/halo-sigs/plugin-page-cache)
  - [Meilisearch 2.0.1](https://github.com/Rainsheep/halo-plugin-meilisearch)
  - [Data Studio（数据工厂） 1.0.0-alpha.2](https://github.com/halo-sigs/plugin-data-studio)
  - [SEO 工具集 1.1.0](https://www.halo.run/store/apps/app-FNGbT)
  - [Metrics Graph 1.0.0](https://github.com/guqing/plugin-metrics-graph)
  - [应用市场 1.5.0](https://www.halo.run/store/apps/app-VYJbF)
  - [Sitemap 1.1.2](https://github.com/halo-dev/plugin-sitemap)
  - [搜索组件 1.6.2](https://github.com/halo-dev/plugin-search-widget)
  - [lightgallery.js 灯箱 1.2.1](https://github.com/halo-sigs/plugin-lightgallery)
  - [highlight.js 代码高亮 1.3.0](https://github.com/halo-sigs/plugin-lightgallery)

### 使用的哪种方式运行？

Docker Compose

### 发生了什么？

Nginx 配置：

<img width="1270" alt="image" src="https://github.com/user-attachments/assets/260ec92c-44ce-4281-bf96-170a1a0c920b">

DEBUG 日志：

```
ec11-4094-828a-fa0df6e7c6df] with result: null, usage: 8
2024-09-10T12:37:13.423+08:00 DEBUG 7 --- [run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler-t-1] r.h.a.c.a.r.LocalThumbnailsReconciler    : Failed to generate thumbnail for: https://blog.fit2cloud.com/upload/2%20%E8%AE%A1%E7%AE%97%E7%BB%93%E6%9E%9C-rxmb.png

java.lang.IllegalStateException: java.io.IOException: Server returned HTTP response code: 403 for URL: https://blog.fit2cloud.com/upload/2%20%E8%AE%A1%E7%AE%97%E7%BB%93%E6%9E%9C-rxmb.png
	at run.halo.app.core.attachment.ThumbnailGenerator.generate(ThumbnailGenerator.java:49) ~[classes/:na]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.lambda$tryGenerateByUrl$1(LocalThumbnailsReconciler.java:123) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.tryGenerate(LocalThumbnailsReconciler.java:112) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.tryGenerateByUrl(LocalThumbnailsReconciler.java:121) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.generateThumbnail(LocalThumbnailsReconciler.java:107) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.requestGenerateThumbnail(LocalThumbnailsReconciler.java:93) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.lambda$reconcile$0(LocalThumbnailsReconciler.java:51) ~[classes/:2.19.2]
	at java.base/java.util.Optional.ifPresent(Unknown Source) ~[na:na]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.reconcile(LocalThumbnailsReconciler.java:49) ~[classes/:2.19.2]
	at run.halo.app.core.attachment.reconciler.LocalThumbnailsReconciler.reconcile(LocalThumbnailsReconciler.java:37) ~[classes/:2.19.2]
	at run.halo.app.extension.controller.DefaultController$Worker.run(DefaultController.java:163) ~[api-2.19.2.jar:na]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
	at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
Caused by: java.io.IOException: Server returned HTTP response code: 403 for URL: https://blog.fit2cloud.com/upload/2%20%E8%AE%A1%E7%AE%97%E7%BB%93%E6%9E%9C-rxmb.png
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source) ~[na:na]
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source) ~[na:na]
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(Unknown Source) ~[na:na]
	at java.base/java.net.URL.openStream(Unknown Source) ~[na:na]
	at run.halo.app.core.attachment.ThumbnailGenerator$ImageDownloader.downloadFileInternal(ThumbnailGenerator.java:175) ~[classes/:na]
	at run.halo.app.core.attachment.ThumbnailGenerator$ImageDownloader.downloadFile(ThumbnailGenerator.java:160) ~[classes/:na]
	at run.halo.app.core.attachment.ThumbnailGenerator.generate(ThumbnailGenerator.java:46) ~[classes/:na]
	... 15 common frames omitted
```

/kind bug
/area core

### 复现步骤

_No response_

### 相关日志输出

_No response_

### 附加信息

_No response_

--- a/application/src/main/java/run/halo/app/core/attachment/ThumbnailGenerator.java
+++ b/application/src/main/java/run/halo/app/core/attachment/ThumbnailGenerator.java
@@ -6,6 +6,7 @@ import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
 import java.net.URISyntaxException;
 import java.net.URL;
@@ -172,7 +173,11 @@ public class ThumbnailGenerator {
             File tempFile = File.createTempFile("halo-image-thumb-", ".tmp");
             long totalBytesDownloaded = 0;
             var tempFilePath = tempFile.toPath();
-            try (InputStream inputStream = url.openStream();
+            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+            connection.setRequestProperty("User-Agent",
+                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)"
+                    + " Chrome/92.0.4515.131 Safari/537.36");
+            try (InputStream inputStream = connection.getInputStream();
                  FileOutputStream outputStream = new FileOutputStream(tempFile)) {
 
                 byte[] buffer = new byte[4096];

--- a/application/src/main/java/run/halo/app/core/attachment/reconciler/AttachmentReconciler.java
+++ b/application/src/main/java/run/halo/app/core/attachment/reconciler/AttachmentReconciler.java
@@ -58,7 +58,7 @@ public class AttachmentReconciler implements Reconciler<Request> {
             var annotations = attachment.getMetadata().getAnnotations();
             if (annotations != null) {
                 attachmentService.getPermalink(attachment)
-                    .map(URI::toString)
+                    .map(URI::toASCIIString)
                     .switchIfEmpty(Mono.fromSupplier(() -> {
                         // Only for back-compatibility
                         return annotations.get(Constant.EXTERNAL_LINK_ANNO_KEY);

--- a/application/src/main/java/run/halo/app/core/attachment/reconciler/LocalThumbnailsReconciler.java
+++ b/application/src/main/java/run/halo/app/core/attachment/reconciler/LocalThumbnailsReconciler.java
@@ -27,6 +27,7 @@ import run.halo.app.core.extension.attachment.Attachment;
 import run.halo.app.core.extension.attachment.Constant;
 import run.halo.app.core.extension.attachment.LocalThumbnail;
 import run.halo.app.extension.ExtensionClient;
+import run.halo.app.extension.ExtensionUtil;
 import run.halo.app.extension.ListOptions;
 import run.halo.app.extension.PageRequestImpl;
 import run.halo.app.extension.controller.Controller;
@@ -47,6 +48,9 @@ public class LocalThumbnailsReconciler implements Reconciler<Reconciler.Request>
     public Result reconcile(Request request) {
         client.fetch(LocalThumbnail.class, request.name())
             .ifPresent(thumbnail -> {
+                if (ExtensionUtil.isDeleted(thumbnail)) {
+                    return;
+                }
                 if (shouldGenerate(thumbnail)) {
                     requestGenerateThumbnail(thumbnail);
                     nullSafeAnnotations(thumbnail).remove(REQUEST_TO_GENERATE_ANNO);

--- a/application/src/test/java/run/halo/app/core/attachment/ThumbnailGeneratorTest.java
+++ b/application/src/test/java/run/halo/app/core/attachment/ThumbnailGeneratorTest.java
@@ -4,11 +4,13 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.HttpURLConnection;
 import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -57,7 +59,9 @@ class ThumbnailGeneratorTest {
             String mockImageData = "fakeImageData";
             InputStream mockInputStream = new ByteArrayInputStream(mockImageData.getBytes());
 
-            doAnswer(invocation -> mockInputStream).when(spyImageUrl).openStream();
+            var urlConnection = mock(HttpURLConnection.class);
+            doAnswer(invocation -> urlConnection).when(spyImageUrl).openConnection();
+            doReturn(mockInputStream).when(urlConnection).getInputStream();
 
             var path = imageDownloader.downloadFileInternal(spyImageUrl);
             assertThat(path).isNotNull();
@@ -81,8 +85,9 @@ class ThumbnailGeneratorTest {
             var fileSizeByte = ThumbnailGenerator.MAX_FILE_SIZE + 10;
             byte[] largeImageData = new byte[fileSizeByte];
             InputStream mockInputStream = new ByteArrayInputStream(largeImageData);
-
-            doReturn(mockInputStream).when(spyImageUrl).openStream();
+            var urlConnection = mock(HttpURLConnection.class);
+            doAnswer(invocation -> urlConnection).when(spyImageUrl).openConnection();
+            doReturn(mockInputStream).when(urlConnection).getInputStream();
             assertThatThrownBy(() -> imageDownloader.downloadFileInternal(spyImageUrl))
                 .isInstanceOf(IOException.class)
                 .hasMessageContaining("File size exceeds the limit");

